examples/pwa.js

```diff
@@ -0,0 +1,321 @@
+'use strict';
+
+var shell;
+
+window.onload = () => {
+  shell = new Shell(window);
+};
+
+
+function log(args) {
+  var var_args = Array.prototype.slice.call(arguments, 0);
+  var_args.unshift('[SHELL]');
+  console/*OK*/.log.apply(console, var_args);
+}
+
+
+class Shell {
+
+  constructor(win) {
+    /** @private @const {!Window} */
+    this.win = win;
+
+    /** @private @const {!AmpViewer} */
+    this.ampViewer_ = new AmpViewer(win,
+        win.document.getElementById('doc-container'));
+
+    /** @private {string} */
+    this.currentPage_ = stripHashMarker(win.location.hash);
+
+    win.addEventListener('popstate', this.handlePopState_.bind(this));
+
+    log('STARTED');
+
+    if (this.currentPage_) {
+      this.navigateTo(this.currentPage_);
+    }
+  }
+
+  /**
+   */
+  handlePopState_() {
+    const newPage = stripHashMarker(this.win.location.hash);
+    log('Pop state: ', newPage, this.currentPage_);
+    if (newPage != this.currentPage_) {
+      this.navigateTo(newPage);
+    }
+  }
+
+  /**
+   * @param {string} path
+   * @return {!Promise}
+   */
+  navigateTo(path) {
+    log('Navigate to: ', path);
+    this.currentPage_ = path;
+
+    // Update URL.
+    const newHash = '#' + path;
```

Comment:
Would be pretty easy to use real paths if you want with a custom action in `server.js` that always serves the same file.